### PR TITLE
DAOS-14181 control: Remove bdev scan cache

### PR DESCRIPTION
**Pull request by tanabarr**
Introduce direct scanning of NVMe devices through dRPC. With this new
interface, when an engine-owned SSD is hotplugged (replaced) details
of the new physical device will be available through dmg storage scan.

Changes applied:
- Remove bdev scan cache used to store static controller details.
- Respect functional boundaries better in unit tests through improved
  decoupling of dependencies.
- Implement dRPC callers and handlers to scan and populate nested NVMe
  structures.
- Simplify decision-making when performing scans.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by tanabarr** [comment]
Wanted to get some early reviews on this change as it is significant. Basically "ripping the Band-Aid off" and "breaking some eggs to make an omelette" come to mind with this. Specifically @knard-intel I've gutted a lot of the previous bdev scan logic and I'm trying to simplify the flow in the redesign, as a result I have removed some of the space-adjustment code that you have implemented and I will retrospectively add that back in after I get the basics working. Any feedback appreciated.

**Comment by tanabarr** [comment]
I will be adding engine side code in a separate Commit.

**Comment by tanabarr** [comment]
With commit 645cba416ed61e475dbbda6464e728550472090c `spdk_nvme_connect` in bio_device.c during smd list is returning nil, probably indicating some concurrent access issue:
```
11/09-22:30:02.09 wolf-226 DAOS[1151239/0/5] bio  DBUG src/bio/bio_device.c:603 alloc_ctrlr_info() fetching 0000:2c:00.0 controller details
11/09-22:30:02.09 wolf-226 DAOS[1151239/0/5] bio  ERR  src/bio/bio_device.c:612 alloc_ctrlr_info() spdk_nvme_connect returned nil
```

File: src/include/daos/drpc_modules.h [hunk 1]
```diff
@@ -33,43 +33,43 @@ enum drpc_sec_agent_method {
 };
 
 enum drpc_mgmt_method {
-	DRPC_METHOD_MGMT_KILL_RANK		= 201,
```
Comment (kjacque) [inline src/include/daos/drpc_modules.h:36]: Looks like just whitespace changes. Mind reverting this file?

File: src/control/server/instance_storage_test.go [hunk 2]
```diff
@@ -1,5 +1,5 @@
 //
-// (C) Copyright 2020-2022 Intel Corporation.
+// (C) Copyright 2020-2023 Intel Corporation.
```
Comment (kjacque) [inline src/control/server/instance_storage_test.go:2]: Copyright seems to be the only change here, better to revert

File: src/control/server/instance_storage_rpc.go [hunk 3]
```diff
@@ -156,3 +161,177 @@ func (ei *EngineInstance) StorageFormatNVMe() (cResults proto.NvmeControllerResu
...
+	}
+
+	return bdevScanToProtoResp(ei.storage.ScanBdevs, req)
+}
+
+func smdQueryEngine(ctx context.Context, engine Engine, pbReq *ctlpb.SmdQueryReq) (*ctlpb.SmdQueryResp_RankResp, error) {
+	ei, ok := engine.(*EngineInstance)
+	if !ok {
+		return nil, errors.New("not EngineInstance")
+	}
```
Comment (kjacque) [inline src/control/server/instance_storage_rpc.go:275]: Converting the interface type instantly to the concrete type feels a bit kludgy to me. Is there a way to update the Engine interface to offer the methods needed here?

File: src/control/server/ctl_smd_rpc.go [hunk 4]
```diff
@@ -44,74 +44,13 @@ func queryRank(reqRank uint32, engineRank ranklist.Rank) bool {
...
-
-		listDevsResp, err := ei.ListSmdDevices(ctx, new(ctlpb.SmdDevReq))
-		if err != nil {
-			return errors.Wrapf(err, "rank %d", engineRank)
-		}
-
-		if len(listDevsResp.Devices) == 0 {
-			rResp.Devices = nil
+		// If rank hasn't been requested, don't include its response.
+		if rResp != nil {
```
Comment (kjacque) [inline src/control/server/ctl_smd_rpc.go:52]: Could rResp == nil even if err == nil?

File: src/control/lib/spdk/src/nvme_control_common.c [hunk 5]
```diff
@@ -428,41 +470,46 @@ _collect(struct ret_t *ret, data_copier copy_data, pci_getter get_pci,
...
-		written = snprintf(ctrlr_tmp->pci_type,
-				   sizeof(ctrlr_tmp->pci_type), "%s",
-				   spdk_pci_device_get_type(pci_dev));
-		if (written >= sizeof(ctrlr_tmp->pci_type)) {
-			rc = -NVMEC_ERR_CHK_SIZE;
-			free(pci_dev);
+
+		pci_type = spdk_pci_device_get_type(pci_dev);
+		free(pci_dev);
+		ctrlr_tmp->pci_type = strndup(pci_type, strlen(pci_type));
```
Comment (kjacque) [inline src/control/lib/spdk/src/nvme_control_common.c:508]: Unsafe to invoke strlen to figure out the length for strndup. Is there a sensible maximum for this string?

File: src/mgmt/srv_query.c [hunk 6]
```diff
@@ -293,14 +294,40 @@ bio_query_dev_list(void *arg)
...
 
+static int
+copy_str2ctrlr(char **dst, const char *src)
+{
+	int len;
+
+	D_ASSERT(src != NULL);
+	D_ASSERT(dst != NULL);
+
+	if ((*dst != NULL) && (strlen(*dst) != 0)) {
```
Comment (kjacque) [inline src/mgmt/srv_query.c:305]: Don't use strlen, use strnlen instead

**Review by kjacque — COMMENTED** [review]
Some early feedback with the understanding that this is still a WIP. A little hard to get through, I wonder if the rename/reorganization of the struct could be done as another PR, landed, and then include only the business logic changes in a second PR.

File: src/bio/bio_config.c [hunk 7]
```diff
@@ -375,7 +377,7 @@ load_vmd_subsystem_config(struct json_config_ctx *ctx, bool *vmd_enabled)
 	int			 rc;
 
 	D_ASSERT(ctx->config_it != NULL);
-	D_ASSERT(vmd_enabled != NULL);
+	D_ASSERT(*vmd_enabled == false);
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/bio/bio_config.c:380]: done

File: src/bio/bio_config.c [hunk 8]
```diff
@@ -615,11 +617,11 @@ check_vmd_status(struct json_config_ctx *ctx, struct spdk_json_val *vmd_ss, bool
 {
 	int	rc;
 
+	D_ASSERT(*vmd_enabled == false);
+
 	if (vmd_ss == NULL)
 		return 0;
 
-	D_ASSERT(vmd_enabled != NULL);
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/bio/bio_config.c:621]: done

File: src/bio/bio_config.c [hunk 9]
```diff
@@ -642,20 +644,22 @@ check_vmd_status(struct json_config_ctx *ctx, struct spdk_json_val *vmd_ss, bool
...
 {
 	struct json_config_ctx	*ctx;
 	struct spdk_json_val	*bdev_ss = NULL;
 	struct spdk_json_val	*vmd_ss = NULL;
-	bool			 vmd_enabled = false;
 	int			 rc = 0;
 
 	D_ASSERT(nvme_conf != NULL);
 	D_ASSERT(opts != NULL);
+	D_ASSERT(*vmd_enabled == false);
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/bio/bio_config.c:662]: done

File: src/bio/bio_xstream.c [hunk 10]
```diff
@@ -990,7 +992,6 @@ init_bio_bdevs(struct bio_xs_context *ctxt)
 			return -DER_EXIST;
 		}
 
-		D_INFO("Resetting LED on dev " DF_UUID " \n", DP_UUID(d_bdev->bb_uuid));
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/bio/bio_xstream.c:993]: bio_led_manage now is a noop in the case that VMD is disabled, as a result this is confusing if we aren't actually going to reset the LED. IMO there is enough logging within the bio_led_manage call stack.

File: src/bio/bio_config.c [hunk 11]
```diff
@@ -931,3 +935,137 @@ bio_read_rpc_srv_settings(const char *nvme_conf, bool *enable, const char **sock
...
+ * Fetch bdev controller parameters from spdk_bdev_dump_info_json output.
+ *
+ * \param[out] *b_info		Device info struct to populate
+ * \param[in]	json		Raw JSON to parse
+ * \param[in]	json_size	Number of JSON chars
+ *
+ * \returns	 Zero on success, negative on failure (DER)
+ */
+int
+bio_decode_bdev_params(struct bio_dev_info *b_info, const void *json, int json_size)
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/bio/bio_config.c:973]: It is called by bio_device.c:json_write_cb which has signature `json_write_cb(void *cb_ctx, const void *json, size_t json_size)`. This is used to write the JSON from spdk_bdev_dump_info_json using spdk_json_write_{begin,end}. So basically its a requirement of the API and I don't feel that casting it within the callback is a good idea.

File: src/bio/bio_config.c [hunk 12]
```diff
@@ -931,3 +935,137 @@ bio_read_rpc_srv_settings(const char *nvme_conf, bool *enable, const char **sock
...
+	struct spdk_json_val *ctrlr_data = NULL;
+	struct spdk_json_val *ns_data    = NULL;
+	void                 *end;
+	ssize_t               values_cnt;
+	char                 *tmp = NULL;
+	ssize_t               rc;
+	char                 *end1 = NULL;
+
+	/* Trim chars to get single valid "nvme" object from array. */
+	tmp = strstr(json, "{");
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/bio/bio_config.c:989]: spdk_bdev_dump_info_json As part of the API should be pretty reliable and not provide corrupted or not-null-terminated strings. I am surprised that it doesn't output enclosing braces. I'm adding the checks as suggested. Done.

File: src/include/daos/drpc_modules.h (hunk 1 above)
Comment (tanabarr) [reply src/include/daos/drpc_modules.h:36]: done

File: src/control/server/instance_storage_test.go (hunk 2 above)
Comment (tanabarr) [reply src/control/server/instance_storage_test.go:2]: done

File: src/control/server/instance_storage_rpc.go (hunk 3 above)
Comment (tanabarr) [reply src/control/server/instance_storage_rpc.go:275]: I have actively been trying to reduce the size of the interface by following the advice in the code comments in the Engine interface definition in harness.go. Ideally we want less methods in the engine interface and not more.

File: src/control/server/ctl_smd_rpc.go (hunk 4 above)
Comment (tanabarr) [reply src/control/server/ctl_smd_rpc.go:52]: Yes, if ranks specified in the request don't match the engine then nil, nil will be returned. This is not behaviour that has been altered in this PR however.

File: src/control/lib/spdk/src/nvme_control_common.c (hunk 5 above)
Comment (tanabarr) [reply src/control/lib/spdk/src/nvme_control_common.c:508]: done

File: src/mgmt/srv_query.c (hunk 6 above)
Comment (tanabarr) [reply src/mgmt/srv_query.c:305]: done

**Comment by tanabarr** [comment]
replaced by https://github.com/daos-stack/daos/pull/13382 and a follow-on PR

**Pull request closed without merging**